### PR TITLE
feat: add tags to a company in intercom

### PR DESCRIPTION
**Pull request by manish339k**
## What are the changes introduced in this PR?

This PR adds a feature in intercom destination. It enables to add tags to a company in group call.

## What is the related Linear task?

Resolves INT-2162

## Please explain the objectives of your changes below

Put down any required details on the broader aspect of your changes. If there are any dependent changes, **mandatorily** mention them here

### Any changes to existing capabilities/behaviour, mention the reason & what are the changes ?

N/A

### Any new dependencies introduced with this change?

N/A

### Any new generic utility introduced or modified. Please explain the changes.

N/A

### Any technical or performance related pointers to consider with the change?

N/A

@coderabbitai review

<hr>

### Developer checklist

- [ ] My code follows the style guidelines of this project

- [ ] **No breaking changes are being introduced.**

- [ ] All related docs linked with the PR?

- [ ] All changes manually tested?

- [ ] Any documentation changes needed with this change?

- [ ] Is the PR limited to 10 file changes?

- [ ] Is the PR limited to one linear task?

- [ ] Are relevant unit and component test-cases added in **new readability format**?

### Reviewer checklist

- [ ] Is the type of change in the PR title appropriate as per the changes?

- [ ] Verified that there are no credentials or confidential data exposed with the changes.

### CODE REVIEW

**Comment by devops-github-rudderstack** [comment]
Test report for this run is available at: https://test-integrations-dev.s3.amazonaws.com/integrations-test-reports/rudder-transformer/3434/test-report.html

File: src/cdk/v2/destinations/intercom/utils.js [hunk 1]
```diff
@@ -369,6 +371,111 @@ const addMetadataToPayload = (payload) => {
...
+        endpoint,
+        finalPayload,
+        {
+          headers,
+        },
+        statTags,
+      );
+      const processedResponse = processAxiosResponse(response);
+      if (!isHttpStatusSuccess(processedResponse.status)) {
+        throw new NetworkError(
```
Comment (shrouti1507) [inline src/cdk/v2/destinations/intercom/utils.js:464]: Every time it should not be network error, if fails for faulty authorization, it is a configuration error,  if company not found, then instrumentation error. Lets take a second look it response parsing and throw error accordingly.

Comment (manish339k) [reply src/cdk/v2/destinations/intercom/utils.js]: made some changes. please check if it looks good.

File: src/cdk/v2/destinations/intercom/utils.js [hunk 2]
```diff
@@ -27,8 +31,30 @@ const {
   SEARCH_CONTACT_ENDPOINT,
   ReservedCompanyProperties,
   CREATE_OR_UPDATE_COMPANY_ENDPOINT,
+  TAGS_ENDPOINT,
 } = require('./config');
 
+const intercomErrorHandler = (message, processedResponse) => {
+  const errorMessages = JSON.stringify(processedResponse?.response?.errors);
```
Comment (utsabc) [inline src/cdk/v2/destinations/intercom/utils.js:38]: keep till `response`

File: src/cdk/v2/destinations/intercom/utils.js [hunk 3]
```diff
@@ -27,8 +31,30 @@ const {
...
+  }
+  if (processedResponse?.status === 401) {
+    throw new ConfigurationError(`${message} : ${errorMessages}`);
+  }
+  if (processedResponse?.status === 404) {
+    throw new InstrumentationError(`${message} : ${errorMessages}`);
+  }
+  throw new NetworkError(
+    `${message} : ${errorMessages}`,
+    processedResponse?.status,
```
Comment (utsabc) [inline src/cdk/v2/destinations/intercom/utils.js:50]: ProcessedResponse is assumed to be present hence no need of optional chaining syntax like `processedResponse?.`

File: src/cdk/v2/destinations/intercom/utils.js [hunk 4]
```diff
@@ -369,6 +396,96 @@ const addMetadataToPayload = (payload) => {
...
+    {
+      headers,
+    },
+    {
+      destType: 'intercom',
+      feature: 'transformation',
+      endpointPath,
+      requestMethod: 'POST',
+      module: 'router',
+    },
```
Comment (utsabc) [inline src/cdk/v2/destinations/intercom/utils.js:428]: Apart from endpointPath can we keep others as `constant` variable and re-use

File: src/cdk/v2/destinations/intercom/utils.js [hunk 5]
```diff
@@ -30,6 +34,27 @@ const {
   TAGS_ENDPOINT,
 } = require('./config');
 
+const intercomErrorHandler = (message, processedResponse) => {
+  const errorMessages = JSON.stringify(processedResponse?.response?.errors);
+  if (processedResponse?.status === 400) {
+    throw new InstrumentationError(`${message} : ${errorMessages}`);
```
Comment (shrouti1507) [inline src/cdk/v2/destinations/intercom/utils.js:40]: Please comment in the code on which error code comes when

File: src/cdk/v2/destinations/intercom/procWorkflow.yaml [hunk 6]
```diff
@@ -208,9 +217,25 @@ steps:
       response.method = "POST";
       response.userId = .message.anonymousId;
       $.context.response.push(response);
+      payload = response.body.JSON;
+      const companyId = await $.createOrUpdateCompany(payload, .destination);
+      $.assert(companyId, "Unable to create or update company");
```
Comment (utsabc) [inline src/cdk/v2/destinations/intercom/procWorkflow.yaml:222]: Why again we are doing this call here?

File: src/cdk/v2/destinations/intercom/procWorkflow.yaml [hunk 7]
```diff
@@ -208,9 +217,25 @@ steps:
...
       const attachUserAndCompanyResponse = $.attachUserAndCompany(.message, .destination.Config);
       attachUserAndCompanyResponse ? attachUserAndCompanyResponse.userId = .message.anonymousId;
       attachUserAndCompanyResponse ? $.context.response.push(attachUserAndCompanyResponse);
+      payload = attachUserAndCompanyResponse.body.JSON;
+      let endpoint = attachUserAndCompanyResponse.endpoint;
+      attachUserAndCompanyResponse ? await $.attachContactToCompany(payload, endpoint, .destination);
+      await $.addOrUpdateTagsToCompany(.message, .destination, companyId);
```
Comment (utsabc) [inline src/cdk/v2/destinations/intercom/procWorkflow.yaml:229]: Same here

File: src/cdk/v2/destinations/intercom/procWorkflow.yaml (hunk 6 above)
Comment (manish339k) [reply src/cdk/v2/destinations/intercom/procWorkflow.yaml:222]: We are doing this call only one time from transformer as we are sending response with statusCode 299. This `groupPayloadForOlderVersion` will only be executed for group call when we select apiVersion 1.4 in UI.

**Review by shrouti1507 — DISMISSED** [review]

File: test/integrations/destinations/intercom/router/data.ts [hunk 8]
```diff
@@ -335,7 +378,51 @@ export const data = [
...
+                    id: '657264e9018c0a647s45',
+                  },
+                  XML: {},
+                  FORM: {},
+                  JSON_ARRAY: {},
+                },
+                endpoint: 'https://api.eu.intercom.io/contacts/70701240741e45d040/companies',
+                files: {},
+                headers: {
+                  Authorization: 'Bearer testApiKey',
```
Comment (utsabc) [inline test/integrations/destinations/intercom/router/data.ts:397]: Please update them to new format of test cases.Take Klaviyo test cases as reference

File: test/integrations/destinations/intercom/router/data.ts [hunk 9]
```diff
@@ -1,187 +1,613 @@
...
+};
+
+export const data: RouterTestData[] = [
   {
+    id: 'Intercom-router-test-1',
+    scenario: 'Framework',
+    successCriteria: 'Some events should be transformed successfully and some should fail',
     name: 'intercom',
-    description: 'Intercom router tests',
+    description: 'Intercom router tests 1',
```
Comment (utsabc) [inline test/integrations/destinations/intercom/router/data.ts:604]: Remove description if its not providing any significance to test case

File: test/integrations/destinations/intercom/router/data.ts [hunk 10]
```diff
@@ -375,222 +773,17 @@ export const data = [
     },
   },
   {
+    id: 'Intercom-router-test-2',
+    scenario: 'Framework',
+    successCriteria: 'Some events should be transformed successfully and some should fail',
```
Comment (utsabc) [inline test/integrations/destinations/intercom/router/data.ts:778]: Whats the difference between this and the previous test case?

File: test/integrations/destinations/intercom/router/data.ts [hunk 11]
```diff
@@ -769,233 +928,27 @@ export const data = [
...
             },
           ],
         },
       },
     },
   },
   {
+    id: 'Intercom-router-test-3',
+    scenario: 'Framework',
+    successCriteria: 'Some events should be transformed successfully and some should fail',
```
Comment (utsabc) [inline test/integrations/destinations/intercom/router/data.ts:943]: Same sucess criteria?

**Review by utsabc — APPROVED** [review]

**Review by anantjain45823 — APPROVED** [review]

**Review by shrouti1507 — APPROVED** [review]

**Pull request merged**